src/rate_limiter/shared_token_bucket_rate_limiter.rs

```diff
@@ -0,0 +1,77 @@
+/// A token bucket rate limiter implementation bases on lazy-update strategy with CPU sharding.
+use crate::{RateLimiter, TokenBucketRateLimiter};
+use std::cmp::min;
+use std::sync::atomic::AtomicUsize;
+use std::sync::atomic::Ordering::Relaxed;
+use std::sync::Arc;
+use std::thread::available_parallelism;
+
+pub struct ShardingTokenBucketRateLimiter {
+    index: AtomicUsize,
+    shards: Vec<Arc<TokenBucketRateLimiter>>,
+}
+
+impl RateLimiter for ShardingTokenBucketRateLimiter {
+    fn acquire(&self) -> Result<(), ()> {
+        let index = self.index.fetch_add(1, Relaxed) % self.shards.len();
+        let limiter = self.shards[index].clone();
+        limiter.acquire()
+    }
+}
+
+impl ShardingTokenBucketRateLimiter {
+    pub fn new(rate_in_seconds: i64, capacity: i64) -> Self {
+        let mut parallelism = available_parallelism().unwrap().get() as i64;
```

Comment:
`available_parallelism` is not ideal choice, since it does not take cgroup into account.
Maybe we should switch to `num_cpus` crate?

Reply:
Ops, I forgot available cpu time in container, I will fix latter

Reply:
I rethought this question. Different from our pursuit of maximum CPU computing performance, the purpose of using shards here is to divide the current limiting target evenly into different buckets and perform current limiting in each shard. The more shards there are, the lower the probability of lock contention. In addition, cgroups provides two controllers, cpu and cpuset. When only cpu is limited but not cpuset, the process can use all CPU cores to consume all cpu quotas in time. Therefore, I think the number of physical CPU cores should be used conservatively.